fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/UnnestedRecordTypeTest.java

```diff
@@ -838,4 +911,378 @@ void indexDoubleNestedType() {
+
+        try (FDBRecordContext context = openContext()) {
+            createOrOpenRecordStore(context, metaData);
+
+            final TestRecordsNestedMapProto.OuterRecord rec = sampleMapRecord();
+            final FDBStoredRecord<Message> storedRecord = recordStore.saveRecord(rec);
+
+            final RecordQuery query = RecordQuery.newBuilder()
+                    .setRecordTypes(List.of(UNNESTED_MAP, TWO_UNNESTED_MAPS))
+                    .setFilter(Query.field(PARENT_CONSTITUENT).matches(Query.field("other_id").equalsValue(rec.getOtherId())))
+                    .build();
+
+            assertNotNull(recordStore.loadRecord(storedRecord.getPrimaryKey()));
+            assertThat(recordStore.executeQuery(query).asList().join(), not(empty()));
+
+            recordStore.deleteRecordsWhere(OUTER, Query.field("other_id").equalsValue(rec.getOtherId()));
+
+            assertNull(recordStore.loadRecord(storedRecord.getPrimaryKey()));
+            assertThat(recordStore.executeQuery(query).asList().join(), empty());
+
+            commit(context);
+        }
+    }
+
+    @Test
+    void deleteWhereFailsWhenParentTypesHaveDifferentNames() {
+        final String secondMapType = "secondMapType";
+        final RecordMetaDataHook hook = setOuterPrimaryKey(concat(recordType(), field("rec_id")))
+                .andThen(addMapType())
+                .andThen(metaDataBuilder -> {
+                    final UnnestedRecordTypeBuilder typeBuilder = metaDataBuilder.addUnnestedRecordType(secondMapType);
+                    typeBuilder.addParentConstituent("p", metaDataBuilder.getRecordType(OUTER));
+                    typeBuilder.addNestedConstituent("map_entry", TestRecordsNestedMapProto.MapRecord.Entry.getDescriptor(), "p", ENTRIES_FAN_OUT);
+                })
+                .andThen(metaDataBuilder -> {
+                    final Index syntheticOtherIndex = new Index("syntheticOther", field("map_entry").nest("key"));
+                    metaDataBuilder.addMultiTypeIndex(List.of(metaDataBuilder.getSyntheticRecordType(UNNESTED_MAP), metaDataBuilder.getSyntheticRecordType(secondMapType)), syntheticOtherIndex);
+                });
+        final RecordMetaData metaData = mapMetaData(hook);
+
+        try (FDBRecordContext context = openContext()) {
+            createOrOpenRecordStore(context, metaData);
+            assertDeleteRecordsWhereFails(OUTER, null, "syntheticOther");
+        }
+    }
+
+    @Test
+    void deleteWhereWithDifferentTypedParents() {
+        // Create a multi-type index on two different unnested types. Each one has a parent constituent named and an inner constituent, so the index is well-defined.
+        // The first column in the index is middle.rec_no, which we also set as the first column in the combined primary key. This means that it should be legal to
+        // perform a delete where with a middle.rec_no predicate
+        final RecordMetaDataHook hook = addMultiTypeDoubleUnnestedIndex()
```

Comment:
I don't have a good answer for this, and any answer would probably be a fair amount of work, but these tests are kind of hard to follow because it's hard to keep track of what things like `addMultiTypeDoubleUnnestedIndex` actually amount to.
I suspect the answer is to either:
1. Have some more larger, models, that have intuitive meanings that we could learn. 
2. Have a more fluent / pithy way of defining metadata so that a test could pretty easily have everything here.

Both of those are fairly sizable bodies of work that require some real thought, so I definitely don't think we should bring them into this PR, but something to keep in mind.

Reply:
Yeah, there's a bit of a trade-off here for brevity and clarity. The way the tests are now, it's actually fairly easy to use the building blocks to create new tests, but only if you already know what the building blocks are. There may be ways with the macros (or macro-like code) to gain back a bit more clarity without sacrificing brevity, but I'm not sure I know how.